### PR TITLE
Fix updating unread marker if not to latest chunk

### PR DESCRIPTION
**Pull request by SpiritCroc**
SetReadMarkerTask was not updating the read marker when both the old and
the new fully read eventId weren't in the last chunk, even when the new
one was after the first one.

Signed-off-by: Tobias Büttner <dev@spiritcroc.de>

 
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Do not assume that the current read marker is in a later chunk then the new requested one, unless it is in the last chunk.

## Tests

1. Open room at permalink in the past
2. Scroll past the unread marker until you are in a later chunk, but not in the last one
3. Exit and re-enter room at permalink in the past
4. Check if read marker line has updated

## Tested devices

- [x] Physical
- [ ] Emulator
- OS version(s): Android 11

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/query/ReadQueries.kt [hunk 1]
```diff
@@ -94,7 +94,7 @@ internal fun isReadMarkerMoreRecent(realmConfiguration: RealmConfiguration,
             val eventToCheckIndex = eventToCheck?.displayIndex ?: Int.MAX_VALUE
             eventToCheckIndex <= readMarkerIndex
         } else {
-            eventToCheckChunk?.isLastForward == false
+            eventToCheckChunk?.isLastForward == false && readMarkerChunk?.isLastForward == true
```
Comment (ganfra) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/query/ReadQueries.kt:97]: Good catch, but I think you should instead check for `readMarkerChunk.isMoreRecentThan(eventToCheckChunk)`

Comment (SpiritCroc) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/query/ReadQueries.kt:97]: Makes sense, thanks!

Comment (SpiritCroc) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/query/ReadQueries.kt:97]: Hm, my first quick test using `eventToCheckChunk != null && readMarkerChunk?.isMoreRecentThan(eventToCheckChunk) == true` doesn't work as intended. Will probably try more tomorrow.

Comment (SpiritCroc) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/query/ReadQueries.kt:97]: Found the issue! There is another bug in `ChunkEntityHelper.isMoreRecentThan()`, pushing the fix as additional commit to this PR.

**Review by ganfra — APPROVED** [review]
LGTM now, thanks!

**Pull request merged**